pvlib/singlediode_methods.py

```diff
@@ -0,0 +1,347 @@
+
+    Returns
+    -------
+    tuple
+        currents [A], voltages [V], power [W], and optionally
+        :math:`\\frac{dI}{dV_d}`, :math:`\\frac{dV}{dV_d}`,
+        :math:`\\frac{dI}{dV}`, :math:`\\frac{dP}{dV}`, and
+        :math:`\\frac{d^2 P}{dV dV_d}`
+    """
+    a = np.exp(diode_voltage / nNsVth)
+    b = 1.0 / resistance_shunt
+    i = photocurrent - saturation_current * (a - 1.0) - diode_voltage * b
+    v = diode_voltage - i * resistance_series
+    retval = (i, v, i*v)
+    if gradients:
+        c = saturation_current * a / nNsVth
+        grad_i = - c - b  # di/dvd
+        grad_v = 1.0 - grad_i * resistance_series  # dv/dvd
+        # dp/dv = d(iv)/dv = v * di/dv + i
+        grad = grad_i / grad_v  # di/dv
+        grad_p = v * grad + i  # dp/dv
+        grad2i = -c / nNsVth  # d2i/dvd
+        grad2v = -grad2i * resistance_series  # d2v/dvd
+        grad2p = (
+            grad_v * grad + v * (grad2i/grad_v - grad_i*grad2v/grad_v**2)
+            + grad_i
+        )  # d2p/dv/dvd
+        retval += (grad_i, grad_v, grad, grad_p, grad2p)
+    return retval
+
+
+def bishop88_i_from_v(voltage, photocurrent, saturation_current,
+                      resistance_series, resistance_shunt, nNsVth,
+                      method='newton'):
+    """
+    Find current given any voltage.
+
+    Parameters
+    ----------
+    voltage : numeric
+        voltage (V) in volts [V]
+    photocurrent : numeric
+        photogenerated current (Iph or IL) in amperes [A]
+    saturation_current : numeric
+        diode dark or saturation current (Io or Isat) in amperes [A]
+    resistance_series : numeric
+        series resistance (Rs) in ohms
+    resistance_shunt : numeric
+        shunt resistance (Rsh) in ohms
+    nNsVth : numeric
+        product of diode ideality factor (n), number of series cells (Ns), and
+        thermal voltage (Vth = k_b * T / q_e) in volts [V]
+    method : str
+        one of two optional search methods: either ``'brentq'``, a reliable and
+        bounded method or ``'newton'`` which is the default.
+
+    Returns
+    -------
+    current : numeric
+        current (I) at the specified voltage (V) in amperes [A]
+    """
+    # collect args
+    args = (photocurrent, saturation_current, resistance_series,
+            resistance_shunt, nNsVth)
+
+    def fv(x, v, *a):
+        # calculate voltage residual given diode voltage "x"
+        return bishop88(x, *a)[1] - v
+
+    if method.lower() == 'brentq':
+        if brentq is NotImplemented:
+            raise ImportError('This function requires scipy')
+        # first bound the search using voc
+        voc_est = estimate_voc(photocurrent, saturation_current, nNsVth)
+        # break out arguments for numpy.vectorize to handle broadcasting
+        vec_fun = np.vectorize(
+            lambda voc, v, iph, isat, rs, rsh, gamma:
+                brentq(fv, 0.0, voc, args=(v, iph, isat, rs, rsh, gamma))
+        )
+        vd = vec_fun(voc_est, voltage, *args)
```

Comment:
The 4 lines above here are confusing to me. Are they equivalent to the code below (which I find more readable)? 

```python
# brentq only works with scalar inputs, so we need a set up a function and
# np.vectorize to repeatedly call the optimizer with the right arguments for
# possible array input
def vd_from_brent(voc, v, iph, isat, rs, rsh, gamma):
    return brentq(fv, 0.0, voc, args=(v, iph, isat, rs, rsh, gamma))
vd_from_brent_vectorized = np.vectorize(vd_from_brent)
vd = vd_from_brent_vectorized(voc_est, voltage, *args)
```

Reply:
✔️